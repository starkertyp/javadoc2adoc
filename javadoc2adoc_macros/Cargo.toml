[package]
name = "javadoc2adoc_macros"
version = "0.1.0"
edition = "2021"

[dependencies]
quote = "1.0.40"
syn = "2.0.100"

[lib]
proc-macro = true

[dev-dependencies]
trybuild = "1.0.104"
