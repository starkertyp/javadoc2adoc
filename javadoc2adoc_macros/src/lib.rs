use proc_macro::TokenStream;
use quote::quote;
use syn::{
    parse::{self, Parser},
    parse_macro_input,
    spanned::Spanned,
    DeriveInput, Error,
};

#[proc_macro_attribute]
pub fn default_javadocable_fields(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as parse::Nothing);
    let mut ast = parse_macro_input!(input as DeriveInput);
    match &mut ast.data {
        syn::Data::Struct(ref mut struct_data) => {
            match &mut struct_data.fields {
                syn::Fields::Named(fields) => {
                    let field = syn::Field::parse_named.parse2(quote! {comment: BlockComment<'a>});
                    match field {
                        Ok(field) => {
                            fields.named.push(field);
                        }
                        Err(error) => {
                            return error.into_compile_error().into();
                        }
                    }

                    let field = syn::Field::parse_named.parse2(quote! {node: Node<'a>});
                    match field {
                        Ok(field) => {
                            fields.named.push(field);
                        }
                        Err(error) => {
                            return error.into_compile_error().into();
                        }
                    }

                    let field = syn::Field::parse_named.parse2(quote! {context: &'a FileContext});
                    match field {
                        Ok(field) => {
                            fields.named.push(field);
                        }
                        Err(error) => {
                            return error.into_compile_error().into();
                        }
                    }
                }
                _ => (),
            }

            return quote! {
                #ast
            }
            .into();
        }
        _ => Error::new(
            ast.span(),
            "default_javadocable_fields is only supported for structs",
        )
        .into_compile_error()
        .into(),
    }
}
