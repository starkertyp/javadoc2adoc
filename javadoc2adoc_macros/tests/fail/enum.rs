use javadoc2adoc_macros::*;

struct Node<'a> {
    thingy: &'a String,
}
struct BlockComment<'a> {
    thingy: &'a String,
}
struct FileContext {}

#[default_javadocable_fields]
enum Test<'a> {}

fn main() {}
