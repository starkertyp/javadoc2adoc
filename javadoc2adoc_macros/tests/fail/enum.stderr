error: default_javadocable_fields is only supported for structs
  --> tests/fail/enum.rs:12:1
   |
12 | enum Test<'a> {}
   | ^^^^
